File: judo-sdk/src/main/java/com/judopay/api/DeDuplicationInterceptor.java
package com.judopay.api;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.judopay.error.DuplicateTransactionError;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import okhttp3.Interceptor;
import okio.Buffer;

class DeDuplicationInterceptor implements Interceptor {

    private static final List<String> UNIQUE_RESPONSES = new ArrayList<>();

    @Override
    public okhttp3.Response intercept(final Interceptor.Chain chain) throws IOException {
        okhttp3.Request request = chain.request();

        if (request.body() != null) {
            String body = bodyToString(request.body());

            JsonParser parser = new JsonParser();
            if (parser.parse(body).isJsonObject()) {
                JsonObject jsonObject = parser.parse(body).getAsJsonObject();

                JsonElement uniqueReference = jsonObject.get("yourPaymentReference");
                JsonElement uniqueRequest = jsonObject.get("uniqueRequest");

                if (uniqueRequest != null && uniqueRequest.getAsBoolean() && uniqueReference != null && UNIQUE_RESPONSES.contains(uniqueReference.getAsString())) {
                    throw new DuplicateTransactionError(uniqueReference.getAsString());
                } else {
                    okhttp3.Response response = chain.proceed(request);

                    if (uniqueReference != null) {
                        UNIQUE_RESPONSES.add(uniqueReference.getAsString());
                    }
                    return response;
                }
            } else {
                return chain.proceed(request);
            }
        } else {
            return chain.proceed(request);
        }
    }

    private String bodyToString(final okhttp3.RequestBody request) throws IOException {
        final Buffer buffer = new Buffer();
        request.writeTo(buffer);
        return buffer.readUtf8();
    }

}

File: judokit-android/src/main/java/com/judokit/android/api/interceptor/DeDuplicationInterceptor.kt
package com.judokit.android.api.interceptor

import com.google.gson.JsonParser
import com.judokit.android.api.error.DuplicateTransactionError
import java.io.IOException
import okhttp3.Interceptor
import okhttp3.RequestBody
import okhttp3.Response
import okio.Buffer

private const val PAYMENT_REFERENCE_KEY = "yourPaymentReference"
private const val UNIQUE_REQUEST_KEY = "uniqueRequest"

internal class DeDuplicationInterceptor : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {

        val request = chain.request()
        val body = request.body()

        if (body != null) {

            val parser = JsonParser()

            val bodAsString = bodyToString(body)
            val jsonElement = parser.parse(bodAsString)

            if (jsonElement.isJsonObject) {
                val jsonObject = jsonElement.asJsonObject
                val uniqueReference = jsonObject[PAYMENT_REFERENCE_KEY]
                val uniqueRequest = jsonObject[UNIQUE_REQUEST_KEY]

                if (uniqueRequest != null &&
                    uniqueRequest.asBoolean &&
                    uniqueReference != null &&
                    UNIQUE_RESPONSES.contains(uniqueReference.asString)
                ) {
                    throw DuplicateTransactionError(uniqueReference.asString)
                } else {
                    val response = chain.proceed(request)
                    if (uniqueReference != null) {
                        UNIQUE_RESPONSES.add(uniqueReference.asString)
                    }
                    return response
                }
            }
        }

        return chain.proceed(request)
    }

    @Throws(IOException::class)
    private fun bodyToString(request: RequestBody): String {
        val buffer = Buffer()
        request.writeTo(buffer)
        return buffer.readUtf8()
    }

    companion object {
        private val UNIQUE_RESPONSES: MutableList<String> = ArrayList()
    }
}
